File: packages/sdk-js/src/core/identity.ts
import { ContractClient } from "../contracts/client";

export class IdentitySkill {
  private client: ContractClient;
  public address: string;

  constructor(client: ContractClient) {
    this.client = client;
    this.address = client.address;
  }

  async ensureRegistered(name: string, niche: string): Promise<boolean> {
    const registered = await this.client.isRegistered(this.address);
    if (registered) return false;
    await this.client.registerAgent(name, niche);
    return true;
  }

  async ensureSkillsAttested(framework = "raw"): Promise<void> {
    const hasIdentity = await this.client.hasSkill(this.address, 0);
    if (!hasIdentity) {
      await this.client.attestAllSkills(framework);
    }
  }

  async getProfile(): Promise<any> {
    const agent = await this.client.getAgent(this.address);
    const skills = await this.client.getAgentSkills(this.address);
    const rep = await this.client.getReputationDisplay(this.address);
    return {
      address: this.address,
      name: agent.name,
      niche: agent.serviceNiche,
      reputation: rep,
      dealsCompleted: agent.dealsCompleted,
      skills,
      isFullyCapable: await this.client.isFullyCapable(this.address),
    };
  }

  async findSellers(
    niche: string, minReputation = 5.0, requireFullSdk = true, limit = 10
  ): Promise<any[]> {
    const candidates = await this.client.getAgentsByNiche(niche, limit * 2);
    const results: any[] = [];
    for (const addr of candidates) {
      const rep = await this.client.getReputationDisplay(addr);
      if (rep < minReputation) continue;
      if (requireFullSdk) {
        const capable = await this.client.isFullyCapable(addr);
        if (!capable) continue;
      }
      const agent = await this.client.getAgent(addr);
      results.push({
        address: addr,
        name: agent.name,
        reputation: rep,
        dealsCompleted: agent.dealsCompleted,
        isFullyCapable: requireFullSdk,
      });
    }
    results.sort((a, b) => b.reputation - a.reputation);
    return results.slice(0, limit);
  }
}

File: packages/sdk-python/anp_sdk/core/identity.py
from ..contracts.client import ContractClient


class IdentitySkill:

    def __init__(self, client: ContractClient):
        self.client = client
        self.address = client.address

    def ensure_registered(self, name: str, niche: str) -> bool:
        if self.client.is_registered(self.address):
            return False
        self.client.register_agent(name, niche)
        return True

    def ensure_skills_attested(self, framework: str = "raw") -> None:
        if not self.client.has_skill(self.address, 0):
            self.client.attest_all_skills(framework)

    def get_profile(self) -> dict:
        agent = self.client.get_agent(self.address)
        skills = self.client.get_agent_skills(self.address)
        rep = self.client.get_reputation_display(self.address)
        return {
            "address": self.address,
            "name": agent.get("name"),
            "niche": agent.get("serviceNiche"),
            "reputation": rep,
            "dealsCompleted": agent.get("dealsCompleted", 0),
            "skills": skills,
            "isFullyCapable": self.client.is_fully_capable(self.address),
        }

    def find_sellers(self, niche: str, min_reputation: float = 5.0,
                     require_full_sdk: bool = True, limit: int = 10) -> list[dict]:
        candidates = self.client.get_agents_by_niche(niche, limit * 2)
        results = []
        for addr in candidates:
            rep = self.client.get_reputation_display(addr)
            if rep < min_reputation:
                continue
            if require_full_sdk and not self.client.is_fully_capable(addr):
                continue
            agent = self.client.get_agent(addr)
            results.append({
                "address": addr,
                "name": agent.get("name"),
                "reputation": rep,
                "dealsCompleted": agent.get("dealsCompleted", 0),
                "isFullyCapable": require_full_sdk,
            })
        results.sort(key=lambda x: x["reputation"], reverse=True)
        return results[:limit]
